app/internalTX.go

```diff
@@ -0,0 +1,197 @@
+			if err != nil {
+				logger.Error("Error in adding to Finalized Queue :", err)
+				return true
+			}
+			transaction.State.Commit()
+		}
+		return false
+	})
+
+	failedProposals := proposals.WithPrefixType(governance.ProposalStateFailed)
+	failedProposals.Iterate(func(id governance.ProposalID, proposal *governance.Proposal) bool {
+		if proposal.Status == governance.ProposalStatusCompleted && proposal.Outcome == governance.ProposalOutcomeCompletedNo {
+			tx, err := GetFinalizeTX(proposal.ProposalID, validator)
+			if err != nil {
+				logger.Error("Error in building TX of type RequestDeliverTx(finalize)", err)
+				return true
+			}
+			err = transaction.AddFinalized(string(proposal.ProposalID), &tx)
+			if err != nil {
+				logger.Error("Error in adding to Finalized Queue :", err)
+				return true
+			}
+			transaction.State.Commit()
+		}
+		return false
+	})
+}
+
+func GetFinalizeTX(proposalId governance.ProposalID, validatorAddress keys.Address) (abciTypes.RequestDeliverTx, error) {
+	finalizeProposal := &gov_action.FinalizeProposal{
+		ProposalID:       proposalId,
+		ValidatorAddress: validatorAddress,
+	}
+
+	txData, err := finalizeProposal.Marshal()
+	if err != nil {
+		return RequestDeliverTx{}, err
+	}
+
+	internalFinalizeTx := abciTypes.RequestDeliverTx{
+		Tx:                   txData,
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     nil,
+		XXX_sizecache:        0,
+	}
+	return internalFinalizeTx, nil
+}
+
+func GetExpireTX(proposalId governance.ProposalID, validatorAddress keys.Address) (abciTypes.RequestDeliverTx, error) {
+	expireVote := &gov_action.ExpireVotes{
+		ProposalID:       proposalId,
+		ValidatorAddress: validatorAddress,
+	}
+
+	txData, err := expireVote.Marshal()
+	if err != nil {
+		return RequestDeliverTx{}, err
+	}
+
+	expireVotes := abciTypes.RequestDeliverTx{
+		Tx:                   txData,
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     nil,
+		XXX_sizecache:        0,
+	}
+	return expireVotes, nil
+}
+
+// Functions for block Ender
+func FinalizeProposals(header *Header, ctx *context, logger *log.Logger) {
+	var finalizeProposals []abciTypes.RequestDeliverTx // Get this from the store
+	ctx.transaction.IterateFinalized(func(key string, tx *abciTypes.RequestDeliverTx) bool {
+		finalizeProposals = append(finalizeProposals, *tx)
+		return false
+	})
+	for _, proposal := range finalizeProposals {
+		actionctx := ctx.Action(header, ctx.deliver)
+		txData := proposal.Tx
+		newFinalize := gov_action.FinalizeProposal{}
+		err := newFinalize.Unmarshal(txData)
+		if err != nil {
+			logger.Error("Unable to UnMarshal TX(Finalize) :", txData)
+			return
+		}
+		uuidNew, _ := uuid.NewUUID()
+		rawTx := action.RawTx{
+			Type: action.PROPOSAL_FINALIZE,
+			Data: txData,
+			Fee:  action.Fee{},
+			Memo: uuidNew.String(),
+		}
+		ok, _ := newFinalize.ProcessDeliver(actionctx, rawTx)
+		if !ok {
+			logger.Error("Failed to Finalize : ", txData, "Error : ", err)
+			return
```

Comment:
You think it might be better to break here? So we can clear the db with the delete

Reply:
I gues the delete , deletes everything ,irrespective of whether deliver was a success or a failure ,but in case it was not a success , it would still be in the passed /failed state , and be picked up again in the next block. 
not breaking would allow us to process the other tx's in the queue